### PR TITLE
Adding a proof

### LINKED ISSUE #12: Ask for inclusion proof
Add a new api that lets a client ask for an inclusion proof of a given key.

### PR DESCRIPTION
**Pull request by ineiti**
- create a proof for a key/value pair in a skipchain
- replace service.GetValue with service.GetProof

In addition, some work on the collection library is done.

Closes #12 

Start by merging this into #26, then merge #26 into master.

### CODE REVIEW

File: omniledger/service/proof.go [hunk 1]
```diff
@@ -0,0 +1,106 @@
...
+
+// ErrorVerifySkipchain is returned if the stored skipblock doesn't
+// have a proper proof that it comes from the genesis block.
+var ErrorVerifySkipchain = errors.New("stored skipblock is not properly evolved from genesis block")
+
+// Verify takes a skipchain id and verifies that the proof is valid for this skipchain.
+// It verifies the collection-proof, that the merkle-root is stored in the skipblock
+// of the proof and the fact that the skipblock is indeed part of the skipchain.
+// If all verifications are correct, the error will be nil.
+func (p Proof) Verify(genesis *skipchain.SkipBlock) error {
```
Comment (kc1212) [inline omniledger/service/proof.go]: Why doesn't Verify do p.Match? I feel it should be a part of this function, the user might forget to call match and think Verify is all that's needed.

File: omniledger/service/proof.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+//   1. Collection proofs the presence or absence of the key. In case of
+//   the key being present, the value is included in the proof
+//   2. Latest is used to verify the merkle tree root used in the collection-proof
+//   is stored in the latest skipblock
+//   3. Links proves that the latest skipblock is part of the skipchain
+//
+// This Structure could later be moved to cothority/skipchain.
+type Proof struct {
+	// Proving that the key/value pair is in the collection.
+	Collection collection.Proof
```
Comment (kc1212) [inline omniledger/service/proof.go]: Maybe rename Collection to InnerProof or something like that?

File: omniledger/service/service.go [hunk 3]
```diff
@@ -217,35 +218,33 @@ func (s *Service) SetKeyValue(req *SetKeyValue) (*SetKeyValueResponse, error) {
...
-
-func getKey(tx *Transaction) []byte {
-	return append(append(tx.Kind, []byte(":")...), tx.Key...)
+	proof, err := NewProof(s.getCollection(req.ID), s.db(), latest.Hash, req.Key)
+	if err != nil {
+		return
+	}
+	resp = &GetProofResponse{
+		Version: CurrentVersion,
+		Proof:   *proof,
```
Comment (kc1212) [inline omniledger/service/service.go]: curious about how proof is serialised. It is the proof from Service, which contains a `collection.Proof`. But `collection.Proof` only contain private fields. Wouldn't verification fail if the proof is sent over the network?

File: omniledger/service/proof.go (hunk 1 above)
Comment (ineiti) [reply omniledger/service/proof.go]: We can have a proof that the key is in there or a proof that the key is missing. A transaction will be one of `Create`, `Update`, `Delete`. So in the case of `Delete` you want to be sure that `Verify` is OK, but that `Match` returns 'false'.

File: omniledger/service/service.go (hunk 3 above)
Comment (ineiti) [reply omniledger/service/service.go]: Good catch! I added an `api_test.go` and created a PR to make this one lighter: #26

**Pull request merged**